flink-connectors/flink-connector-aws/src/main/java/org/apache/flink/streaming/connectors/kinesis/unified/KinesisDataStreamsSinkWriter.java

```diff
@@ -0,0 +1,218 @@
+    private final boolean failOnError;
+
+    private static final Logger LOG = LoggerFactory.getLogger(KinesisDataStreamsSinkWriter.class);
+
+    KinesisDataStreamsSinkWriter(
+            ElementConverter<InputT, PutRecordsRequestEntry> elementConverter,
+            Sink.InitContext context,
+            int maxBatchSize,
+            int maxInFlightRequests,
+            int maxBufferedRequests,
+            long flushOnBufferSizeInBytes,
+            long maxTimeInBufferMS,
+            boolean failOnError,
+            String streamName,
+            Properties kinesisClientProperties) {
+        super(
+                elementConverter,
+                context,
+                maxBatchSize,
+                maxInFlightRequests,
+                maxBufferedRequests,
+                flushOnBufferSizeInBytes,
+                maxTimeInBufferMS);
+        this.failOnError = failOnError;
+        this.streamName = streamName;
+        this.metrics = context.metricGroup();
+        initMetricsGroup();
+        client = buildClient(kinesisClientProperties);
+    }
+
+    private KinesisAsyncClient buildClient(Properties kinesisClientProperties) {
+        final ClientConfiguration clientConfiguration =
+                new ClientConfigurationFactory().getConfig();
+        clientConfiguration.setUseTcpKeepAlive(true);
+
+        final SdkAsyncHttpClient httpClient =
+                AwsV2Util.createHttpClient(
+                        clientConfiguration,
+                        NettyNioAsyncHttpClient.builder(),
+                        kinesisClientProperties);
+
+        return AwsV2Util.createKinesisAsyncClient(
+                kinesisClientProperties, clientConfiguration, httpClient);
+    }
+
+    @Override
+    protected void submitRequestEntries(
+            List<PutRecordsRequestEntry> requestEntries,
+            Consumer<Collection<PutRecordsRequestEntry>> requestResult,
+            Consumer<Exception> exceptionConsumer) {
+
+        PutRecordsRequest batchRequest =
+                PutRecordsRequest.builder().records(requestEntries).streamName(streamName).build();
+
+        LOG.trace("Request to submit {} entries to KDS using KDS Sink.", requestEntries.size());
+
+        CompletableFuture<PutRecordsResponse> future = client.putRecords(batchRequest);
+
+        future.whenComplete(
+                (response, err) -> {
+                    if (err != null) {
+                        LOG.warn(
+                                "KDS Sink failed to persist {} entries to KDS",
+                                requestEntries.size(),
+                                err);
+                        totalFullyFailedFlushesCounter.inc();
+                        numRecordsOutErrorsCounter.inc(requestEntries.size());
+
+                        if (isRetryable(err, exceptionConsumer)) {
+                            requestResult.accept(requestEntries);
+                        }
```

Comment:
Is it enough to only retry completely failed requests or should we also retry the elements from a partially failed one?

Reply:
So I retry elements from a partially failed one on line 178, where I build up the elements that have failed in a batch where some have succeeded and add those back to the buffer.